File: server/app/schemas/match.py
from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


class MatchResponse(BaseModel):
    id: int
    message_id: int
    interest_id: int
    preco_encontrado: float | None
    score: float
    raw_text_snippet: str | None
    matched_keyword: str | None
    llm_motivo: str | None
    llm_aprovado: bool
    llm_validado: bool
    preco_ok: bool
    alerted: bool
    alerted_at: datetime | None
    lido: bool
    lido_em: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


class MatchDetailResponse(MatchResponse):
    chat_name: str | None
    message_text: str | None
    message_link: str | None
    produto_nome: str


class MatchReadResponse(BaseModel):
    id: int
    lido: bool
    lido_em: datetime | None


class MatchBulkReadResponse(BaseModel):
    updated: int


class MatchPeriod(str, Enum):
    hoje = "hoje"
    d7 = "7d"
    d30 = "30d"
    tudo = "tudo"


class MatchStatus(str, Enum):
    alertado = "alertado"
    reprovado = "reprovado"


class MatchOrderBy(str, Enum):
    data = "data"
    preco = "preco"
    score = "score"


class MatchOrderDir(str, Enum):
    asc = "asc"
    desc = "desc"


class MatchFilterParams(BaseModel):
    """Filtros do feed de matches.

    Usado nos dois lados: como query-param model no `GET /matches` e como body
    do `POST /matches/read-all`. Um schema só é o que garante que "marcar todos
    como lidos" opere exatamente sobre o que está na tela.

    Sobre `status`: `nao_lidos` é o eixo de leitura e faz AND. `alertado` e
    `reprovado` são o eixo de resultado e fazem OR **entre si** — se todos
    fizessem AND, "alertado + reprovado" nunca retornaria nada, já que um match
    reprovado pela IA jamais é alertado.
    """

    periodo: MatchPeriod = MatchPeriod.tudo
    tz: str | None = None
    nao_lidos: bool = False
    status: list[MatchStatus] = Field(default_factory=list)
    chat_id: int | None = None
    preco_min: float | None = Field(None, ge=0)
    preco_max: float | None = Field(None, ge=0)
    order_by: MatchOrderBy = MatchOrderBy.data
    order_dir: MatchOrderDir = MatchOrderDir.desc


class MatchListQuery(MatchFilterParams):
    """Filtros + paginação para o `GET /matches`.

    Existe separado porque o FastAPI não aceita um query-param model convivendo
    com parâmetros `Query()` avulsos na mesma assinatura — ou tudo entra no
    modelo, ou a expansão do modelo não acontece. Herdar mantém o body do
    `read-all` (que é `MatchFilterParams`) sem campos de paginação.
    """

    skip: int = Field(0, ge=0)
    limit: int = Field(30, ge=1, le=200)


class MatchListResponse(BaseModel):
    items: list[MatchDetailResponse]
    total: int
    has_more: bool


class MatchStatsResponse(BaseModel):
    nao_lidos: int
    novos_hoje: int
    ultimas_24h: int
    ultimos_7d: int
    interesses_ativos: int


class MatchChatResponse(BaseModel):
    chat_id: int
    chat_name: str | None
    total: int

File: web/src/types/match.ts
export interface MatchDetailResponse {
  id: number
  message_id: number
  interest_id: number
  preco_encontrado: number | null
  score: number
  raw_text_snippet: string | null
  matched_keyword: string | null
  llm_motivo: string | null
  llm_aprovado: boolean
  llm_validado: boolean
  preco_ok: boolean
  alerted: boolean
  alerted_at: string | null
  lido: boolean
  lido_em: string | null
  created_at: string
  chat_name: string | null
  message_text: string | null
  message_link: string | null
  produto_nome: string
}

export type MatchPeriod = 'hoje' | '7d' | '30d' | 'tudo'
export type MatchStatus = 'alertado' | 'reprovado'
export type MatchOrderBy = 'data' | 'preco' | 'score'
export type MatchOrderDir = 'asc' | 'desc'

/**
 * Filtros do feed. `nao_lidos` combina em AND com o resto; `alertado` e
 * `reprovado` combinam em OR entre si (ver o schema equivalente no backend).
 */
export interface MatchFilters {
  periodo: MatchPeriod
  status: MatchStatus[]
  nao_lidos: boolean
  chat_id?: number
  preco_min?: number
  preco_max?: number
  order_by: MatchOrderBy
  order_dir: MatchOrderDir
}

export const DEFAULT_MATCH_FILTERS: MatchFilters = {
  periodo: '7d',
  status: [],
  nao_lidos: false,
  order_by: 'data',
  order_dir: 'desc',
}

export interface MatchListResponse {
  items: MatchDetailResponse[]
  total: number
  has_more: boolean
}

export interface MatchStatsResponse {
  nao_lidos: number
  novos_hoje: number
  ultimas_24h: number
  ultimos_7d: number
  interesses_ativos: number
}

export interface MatchChatResponse {
  chat_id: number
  chat_name: string | null
  total: number
}

export interface MatchReadResponse {
  id: number
  lido: boolean
  lido_em: string | null
}
